operations/secure_random/secure_random.cc

```diff
@@ -0,0 +1,203 @@
+    free(buf_);
+    free(extra_block_);
+  }
+
+  void GenerateData(int minval, int maxval) {    
+    auto flat = output->flat<T>();
+
+    randombytes_buf_deterministic(buf_, bytes_count_, seeds);
+
+    Uniform(minval, maxval);
+
+    std::copy(buf_, buf_ + flat.size(), flat.data());
+  }
+
+private:
+  T *buf_ = NULL;
+  int count_ = 0;
+  int bytes_count_ = 0;
+
+  T * extra_block_ = NULL;
+  uint32_t block_counter_ = 0;
+  int elements_per_block_ = 0;
+  int inner_block_index_ = 0;
+
+  void Uniform(T lo, T hi) {
+    auto range = static_cast<typename std::make_unsigned<T>::type>(hi) -
+                   static_cast<typename std::make_unsigned<T>::type>(lo);
+
+    typename std::make_unsigned<T>::type min = (1U + ~hi) % hi;
+
+    for (int i = 0; i < count_; ++i) {
+      auto unsign = static_cast<typename std::make_unsigned<T>::type>(buf_[i]);
+        while(unsign < min) {
+          // rejection sampling, get the next valid number in the stream
+          buf_[i] = GetNextValidData();
+          unsign = static_cast<typename std::make_unsigned<T>::type>(buf_[i]);
+        }
+        buf_[i] = random::SignedAdd(lo, buf_[i] % range);
+    }
+  }
+
+  T GetNextValidData() {
+    // if the extra block has been used up get the next available block
+    if(inner_block_index_ + 1 == elements_per_block_) {
+      inner_block_index_ = 0;
+      block_counter_++;
+
+      randombytes_buf_deterministic_ic(extra_block_, CHACHABLOCKSIZE, block_counter_, seeds);
+    }
+
+    T ret = extra_block_[inner_block_index_];
+    inner_block_index_++;
+
+    return ret;
+  }
+};
+
+template <typename T>
+class SecureRandomOp : public OpKernel {
+public:
+  explicit SecureRandomOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
+    const Tensor& shape_t = context->input(0);
+    const Tensor& seed_t = context->input(1);
+    const Tensor& minval = context->input(2);
+    const Tensor& maxval = context->input(3);
+    TensorShape shape;
+    OP_REQUIRES_OK(context, MakeShape(shape_t, &shape));
+    OP_REQUIRES(context, seed_t.dims() == 1 && seed_t.dim_size(0) == 8,
+                errors::InvalidArgument("seed must have shape [8], not ",
+                                        seed_t.shape().DebugString()));
+
+    OP_REQUIRES(context, TensorShapeUtils::IsScalar(maxval.shape()),
+                errors::InvalidArgument("maxval must be 0-D, got shape ",
+                                        maxval.shape().DebugString()));
+
+    T hi = maxval.scalar<T>()();
+    T lo = minval.scalar<T>()();
+    OP_REQUIRES(
+      context, lo < hi,
+      errors::InvalidArgument("Need minval < maxval, got ", lo, " >= ", hi));
+
+    // Allocate output
+    Tensor* output;
+    OP_REQUIRES_OK(context, context->allocate_output(0, shape, &output));
+    if (shape.num_elements() == 0) return;
+
+    if (sodium_init() < 0) {
+      return;
+    }
+
+    int number_of_seeds = randombytes_SEEDBYTES / sizeof(int32);
```

Comment:
where does `randombytes_SEEDBYTES` come from, and does it related to the `8` earlier?

Reply:
Yes. Its 32 bytes. It comes from the libsodium header file.

Reply:
any way it could be unified (eg as a `#define`) so we don't have magic numbers (`8`)?